autoprotocol/protocol.py

```diff
@@ -888,8 +889,14 @@ def as_dict(self):
             for a in dir(self)
             if not a.startswith("__") and not callable(getattr(self, a))
         ]
-
-        explicit_props = ["outs", "refs", "instructions", "time_constraints"]
+        # if either refs or instructions is empty raise error
+        if not self.refs or not self.instructions:
+            raise RuntimeError("there must be at least one instruction and a ref")
+        # time_constraints is an optional attribute and should not be serialized if value is []
+        if not self.time_constraints:
+            explicit_props = ["outs", "refs", "instructions"]
+        else:
+            explicit_props = ["outs", "refs", "instructions", "time_constraints"]
```

Comment:
@yangchoo Added check to exclude 'time_constraints' if empty. 
Also added RuntimeError if either refs or instructions is empty, but did we want this in the end? Or did you want to just serialize []? (Wasn't 100% sure, now that time_constraints is set to []...)

Reply:
oof. Yea... this is ugly. 

I think this is fine. I would clean this up a little, by making this distinction clearer between things we always serialize even when empty or not

E.g.
```
explicit_props = ["outs", "refs", "instructions"]  # things we always serialize
optional_props = ["time_constraints"]  # things we serialize if it is not empty

for prop in optional_props:
  if not prop:
    explicit_prop.append(prop)
```